File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        
        unordered_map<Node*, Node*> mp;
        
        Node* curr = head;
        Node* dummy = NULL;
        
        while(curr != NULL){
            
            dummy = new Node(curr->val);
            mp[curr] = dummy;
            curr = curr->next;
        }
        
        curr = head;
        while(curr != NULL){
            
            dummy = mp[curr];
            dummy->next = mp[curr->next];
            dummy->random = mp[curr->random];
            curr = curr->next;
        }
        
        return mp[head];
        
    }
};

File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        
        hashmap = {None:None}
        curr = head
        dummy = None
        
        while curr:
            dummy = Node(curr.val, curr.next, curr.random)
            hashmap[curr] = dummy
            curr = curr.next
        
        curr = head
        while curr:
            dummy = hashmap[curr]
            dummy.next = hashmap[curr.next]
            dummy.random = hashmap[curr.random]
            curr = curr.next
        return hashmap[head]
    